### PR TITLE
Publish Landmark markers for RViz.

### PR DESCRIPTION
**Pull request by pifon2a**
[RFC=0011](https://github.com/googlecartographer/rfcs/blob/master/text/0011-landmarks.md)

### CODE REVIEW

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc [hunk 1]
```diff
@@ -55,6 +56,32 @@ visualization_msgs::Marker CreateTrajectoryMarker(const int trajectory_id,
   return marker;
 }
 
+int GetLandmarkIndex(const std::string& landmark_id) {
+  static auto* landmark_id_to_index = new std::unordered_map<std::string, int>;
```
Comment (cschuet) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Ehm...

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc [hunk 2]
```diff
@@ -55,6 +56,32 @@ visualization_msgs::Marker CreateTrajectoryMarker(const int trajectory_id,
...
+  }
+  return it->second;
+}
+
+visualization_msgs::Marker CreateLandmarkMarker(const std::string& landmark_id,
+                                                const Rigid3d& landmark_pose,
+                                                const std::string& frame_id) {
+  visualization_msgs::Marker marker;
+  marker.ns = "Landmarks";
+  int landmark_index = GetLandmarkIndex(landmark_id);
```
Comment (cschuet) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Why not just hash the landmark_id?

Comment (pifon2a) [reply cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Hashes can have collisions.

Comment (ojura) [reply cartographer_ros/cartographer_ros/map_builder_bridge.cc]: I think that using a string for an ID doesn't mesh well with the rest of the cartographer codebase - all other IDs (submap, node) are integer-based.

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc [hunk 3]
```diff
@@ -55,6 +56,32 @@ visualization_msgs::Marker CreateTrajectoryMarker(const int trajectory_id,
   return marker;
 }
 
+int GetLandmarkIndex(
+    const std::string& landmark_id,
+    std::unordered_map<std::string, int>* landmark_id_to_index) {
+  auto it = landmark_id_to_index->find(landmark_id);
```
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Why not just use

```c++
const int new_index = landmark_id_to_index->size()++;
landmark_id_to_index->emplace(std::make_pair(landmark_id, new_index));
return landmark_id_to_index->at(landmark_index);
```
Should get rid of the if condition.
http://en.cppreference.com/w/cpp/container/unordered_map/emplace

(not tested)

Comment (pifon2a) [reply cartographer_ros/cartographer_ros/map_builder_bridge.cc]: I think it is less readable.

**Pull request merged**